gateway/middleware.go

```diff
@@ -381,21 +380,46 @@ func (t BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
 	t.clearSession(session)
 
 	didQuota, didRateLimit, didACL, didComplexity := make(map[string]bool), make(map[string]bool), make(map[string]bool), make(map[string]bool)
-	policies := session.PolicyIDs()
 
-	for _, polID := range policies {
-		t.Gw.policiesMu.RLock()
-		policy, ok := t.Gw.policiesByID[polID]
-		t.Gw.policiesMu.RUnlock()
-		if !ok {
-			err := fmt.Errorf("policy not found: %q", polID)
-			t.Logger().Error(err)
-			if len(policies) > 1 {
-				continue
+	var (
+		policyIDs         = session.PolicyIDs()
+		customPolicies    []user.Policy
+		useCustomPolicies bool
+		err               error
+	)
+
+	if customPolicies, err = session.CustomPolicies(); err != nil {
+		t.Logger().Error(err)
+	} else {
+		if len(customPolicies) > 0 {
+			useCustomPolicies = true
+			policyIDs = []string{}
+			for pi, _ := range customPolicies {
+				policyIDs = append(policyIDs, fmt.Sprintf("%d", pi))
 			}
+		}
+	}
 
-			return err
+	for _, polID := range policyIDs {
+		var policy user.Policy
+
+		if useCustomPolicies {
+			polIdx, _ := strconv.ParseInt(polID, 10, 64)
+			policy = customPolicies[polIdx]
```

Comment:
Unhandled err above puts into question if `customPolicies` index is out of range.

Reply:
error handled

Reply:
@padiazg i think you missed a push there (expected a length check or a return on L392)